### DISCUSSION TITLE
Has anyone investigated supporting the CUDA version updates that are enabled by jetpack5?

### DISCUSSION THREAD

**Discussion by brt-brennan-coslett** [OE4T/meta-tegra#1275] [Q&A]
It looks like starting w/ Jetpack5 NVIDIA supports these CUDA version update packages [see Linux/aarch64-jetson here](https://developer.nvidia.com/cuda-11-8-0-download-archive?target_os=Linux&target_arch=aarch64-jetson&Compilation=Native&Distribution=Ubuntu&target_version=20.04&target_type=deb_local) for their stock Ubuntu rootfs and see also [this article on their website](https://developer.nvidia.com/blog/simplifying-cuda-upgrades-for-nvidia-jetson-users/).

It looks like the local deb repo has all of the needed cuda debs for supporting this but that meta-tegra would need some additional work to support that. Is that something thats planned or planned to be investigated in the future?

**Comment by ichergui** [comment]
Hey @brennan-brt 
Welcome to the OE4T community.
We are not planning to get the latest CUDA in `meta-tegra`. I will discuss with @madisongh  and we will share with you any update.
Could you please share when you need that ?

**Reply by brt-brennan-coslett** [reply]
We were looking at it for the August-September timeframe. I was probably going to try to get it working for us locally sometime around then but hooking into the shared-cuda-libraries is some of the more complicated stuff in meta-tegra. So if that is possibly in your guys timeline that would be amazing. Thank you for the quick response!

**Comment by madisongh** [comment]
I haven't really had time to look at this in depth, but based on what I see, it should be possible to implement.  The recipes themselves shouldn't be too hard to put in place, reusing the `.inc` files we already have.  The somewhat tricky part is that these upgrades are *not* drop-in replacements for the base CUDA support provided in the main JetPack SDK, but are *additions* -- the originals still have to be present to support the other JetPack packages that use CUDA (TensorRT, cuDNN, VPI, etc.).  That's what it looks like to me, at least.

So it would be a matter of (using 11.8 as an example):
* Making copies of the existing `cuda-*` recipes, using a base name that includes the updated version, e.g. `cuda-11-8-`, and updating them to use the `cuda/compute/<version>` feed, update the SHAs, and account for any content changes, etc.
* Adding a `cuda-11-8-compat` recipe for installing the "driver" libraries from the `cuda-compat` deb package

With that in place, you should be able then create recipes that `inherit cuda`, followed by overrides of some of the variables:
* Setting `CUDA_VERSION` to `11.8`
* Setting `CUDA_DEPENDS`  and `CUDA_NATIVEDEPS` to depend on the version-prefixed recipes (e.g. `cuda-11-8-libraries`)
* Maybe create a wrapper script for invoking your application to set `LD_LIBRARY_PATH` to point to the directory where the updated driver libraries get installed.

There might be some changes we could make in the layer to make this a bit easier/automatic, but I'd wait on those until some progress has been made on an initial implementation.  I'm also not exactly sure how to make this play well with SDK creation, to allow SDK users to use any of the installed CUDA versions for their builds, but that's probably not too hard to work out.

**Comment by ichergui** [comment]
Hello guys, 
As promised, I started working on this. Please check the WIP PR #1282

**Reply by brt-brennan-coslett** [reply]
Awesome, thanks again for working on this so quickly! I wasn't sure if it was better to leave thoughts in this chain or on the PR, lmk if I should move this to the PR. 

A couple questions/thoughts I had from looking over the PR:
- Is the `NVIDIA_DEVNET_MIRROR` config change in your local.conf a requirement for using a different cuda runtime version or just for speeding up the WIP?
- I think for the final implementation it would be really nice to have a `CUDA_RUNTIME_VERSION` value that you could set local or distro conf and that would get used something like this to do the PREFERRED_VERSION configs you have in the local.conf right now.
```
import cuda-runtime-${CUDA_RUNTIME_VERSION}.inc
```

**Reply by ichergui** [reply]
Hey @brennan-brt 
The `NVIDIA_DEVNET_MIRROR` is used to speed up things and validate the new `CUDA` recipes, I need to add the `CUDA` driver 11.8. 
I will continue working on the PR hopefully no need for `NVIDIA_DEVNET_MIRROR` in the next updates 
I will keep you updated

**Comment by ichergui** [comment]
Hey @brennan-brt 

PR #1282  is ready for final review and test. 
Could you please try it and share your experience with it ?

**Reply by brennancoslett** [reply]
Should there be another commit in the PR or a force-push? I still just see the original. Thanks!

**Reply by ichergui** [reply]
I just used force-push method  
You should see something different then previous PR.
Please follow the instructions mentioned in the PR for the testing.

**Reply by brennancoslett** [reply]
I see it now, will try to check it out tomorrow.

**Reply by elbbit01** [reply]
Hi guys, great job here.

I have trying to reproduce the steps for my use case. My system is still based on kirkstone, and the hardware is Jetson Xavier AGX, so it needs jetson Linux 35.X.
We read that 35.X should still be compatible with cuda 12.2 and this is actually required by some of our developers, so I am trying to update to cuda 12.2.

If I check master or other branches with Jetson 36.X, I see some of the cuda-* files have version 12.6, but some others older. I fail to find a reference to know which versions to update and which not. Can anybody give me a hint? Where can I find the proper dependency versions to build with CUDA_VERSION = 12.2? Is this possible at all? 

I have also tried to just backport the versions of Jetson 36.X and see if they still work, but I got into a bigger dependency hell and I think I need to step back and try something simpler.

Thanks in advance,
Alvaro.
